[proxy] More flexible container name -> hostname derivation

2opremio opened the pull request:
Provides a new "--hostname=substitution" flag to WeaveProxy, where
'substitution' is a sed-style substitution command. 'substitution' will be
applied to container names at launch-time, as a means to control what hostnames will
be registered in WeaveDNS.

The substitution is implemented with
https://golang.org/pkg/regexp/#Regexp.ReplaceAllString . Thus, references to
regexp substitution groups should be prepended with a dollar sign.

For instance, if we provide "--hostname='/aws-[0-9]+-(.*)/my-app-$1/'", running
a container named 'aws-12798186823-foo' through WeaveProxy will lead to WeaveDNS
registering 'my-app-foo'.

Closes #1018

tomwilkie commented:
we'll need a few lines in the docs too, please.

2opremio commented:
@tomwilkie Sure, I considered it but it wasn't clear to me were to include it.  https://github.com/weaveworks/weave/tree/master/docs only seems to include architectural stuff.

2opremio commented:
OK, I found `/site` which I hadn't payed attention to until now. I will add something there first thing tomorrow.

2opremio commented:
@tomwilkie Done, I will squash the commit once the PR is good to go.

2opremio commented:
Rebased with master (and squashed the documentation commit) to make coveralls happy

tomwilkie commented:
You should fetch the coverage.txt from your branch and diff it with master to see where this -5% is coming from.

rade commented on weave:
```diff
@@ -36,7 +36,7 @@ weave launch-router [--password <password>] [--nickname <nickname>]
 weave launch-dns    [<addr>]
 weave launch-proxy  [-H <endpoint>] [--with-dns | --without-dns]
-                      [--no-default-ipalloc] [--no-rewrite-hosts]
+                      [--no-default-ipalloc] [--no-rewrite-hosts] [--hostname]
```
That should be `--hostname <figureoutwhattoputhere>`

2opremio commented:
@tomwilkie Thanks, will do. I suspect it's coming from `proxy/` not having any tests until now.

rade commented on proxy/proxy.go:
```diff
@@ -15,20 +16,25 @@ import (
-	execCreateRegexp      = regexp.MustCompile("^(/v[0-9\\.]*)?/containers/[^/]*/exec$")
+	containerCreateRegexp = regexp.MustCompile(`^(/v[0-9\.]*)?/containers/create$`)
+	containerStartRegexp  = regexp.MustCompile(`^(/v[0-9\.]*)?/containers/[^/]*/(re)?start$`)
+	execCreateRegexp      = regexp.MustCompile(`^(/v[0-9\.]*)?/containers/[^/]*/exec$`)
```
so those regexps were wrong previously? e.g. they would match `/v\/foo`? If so, raise a bug and fix on the 1.0 branch.

2opremio replied:
I don't think there were wrong, I just wanted to use backquotes for all of them (for uniformity). Maybe it's my lack of golang knowledge, but aren't they equivalent? (as I understand backquotes don't require escaping) http://play.golang.org/p/SUiW2R0B_P

rade replied:
Ah, I completely missed quote->backquote change.

rade commented on proxy/proxy_test.go:
```diff
@@ -0,0 +1,59 @@
+		t.Error("Unexpected replacement hostname", hostname)
+	}
+
+}
```
use `testify`'s `require`, as we do elsewhere in our unit tests.

2opremio replied:
Done

rade commented on proxy/proxy.go:
```diff
@@ -119,6 +132,31 @@ func (proxy *Proxy) ListenAndServe() {
+	escapedPattern := namedSubmatches[substitutionPatternName]
+	escapedReplacement := namedSubmatches[substitutionReplacementName]
+	unescapedPattern := unescapeBackSlashes(escapedPattern)
+	unescapedReplacement := unescapeBackSlashes(escapedReplacement)
```
I'd just make this
````
pattern := unescapeBackSlashes(namedSubmatches[substitutionPatternName])
replacement :=  unescapeBackSlashes(namedSubmatches[substitutionReplacementName])
````

2opremio replied:
I, at the very least,  need `escapedPattern` to report an error if `unescapedPattern` doesn't compile (read below).

rade replied:
fair enough; leave as is then.

paulbellamy commented on proxy/proxy.go:
```diff
@@ -119,6 +132,31 @@ func (proxy *Proxy) ListenAndServe() {
+
+	patternRegexp, err := regexp.Compile(unescapedPattern)
+	if err != nil {
+		errStr := fmt.Sprintf("Incorrect hostname pattern '%s': %s", escapedPattern, err.Error())
```
You probably mean `hostname` not `escapedPattern` here.

2opremio replied:
My intention is to show `escapedPattern` to the user, which is the specific part of the substitution command which failed to compile. I could show the full hostname, but `Compile`'s error will refer exclusively to the pattern.

Thanks about `fmt.Errorf`, I will use it instead.

tomwilkie commented:
Alternatively, instead of requireing a sed-style regex in the for /foo/bar/, how about we just have --hostname-match=foo and --hostname-replacement=bar?  It would save us a bunch of code (using regexes to parse regexes...)  Default for --hostname-match should be (.*) and --hostname-replacement should be $1

Also, I wouldn't bother including this in the weave usage text; this feature is meant for integrators only.  Just include it in the site/ docs.

2opremio commented:
@tomwilkie I personally like the `/regexp/substitution/` better, @rade also seems to agree https://github.com/weaveworks/weave/issues/1018 

Also, I am not sure it makes sense not to document it in usage if it's in the docs. If we want it to be hidden from normal users then it shouldn't appear in the site.

tomwilkie commented:
@rade what do you think?  I suggested this as I'm not sure the parsing of the sed-style regexps is worth the extra code.

2opremio commented:
I pushed a commit which addressed all the code comments (except for @tomwikie 's  design remark). Also, I pushed a change to surround relevant regexps with `^` and `$`. 

Please make sure to squash the changes before merging.

rade commented on test/680_proxy_hostname_derivation.sh:
```diff
@@ -0,0 +1,36 @@
+
+check() {
+  assert "proxy exec_on $HOST1 $1 getent hosts $2 | tr -s ' '" "$3 $2"
+}
```
I've just merged #1050, which extends `assert_dns_a_record` to do what you want here.

2opremio replied:
Will do

rade commented:
>  I'm not sure the parsing of the sed-style regexps is worth the extra code.

Agreed, and I am not sold on the UI benefits either.

So two options it is, with the defaults as you suggest. But I do think they should be documented in the usage.

2opremio commented:
OK, I will rework the PR to use those two options.

2opremio commented:
@rade @paulbellamy @tomwilkie please review

paulbellamy commented on proxy/proxy.go:
```diff
@@ -15,11 +15,13 @@ import (
+	dockerSock                  = "/var/run/docker.sock"
+	dockerSockUnix              = "unix://" + dockerSock
+	substitutionPatternName     = "pattern"
+	substitutionReplacementName = "replacement"
```
These are unused now

rade commented on weave:
```diff
@@ -37,6 +37,7 @@ weave launch-router [--password <password>] [--nickname <nickname>]
 weave launch-dns    [<addr>]
 weave launch-proxy  [-H <endpoint>] [--with-dns | --without-dns]
                       [--no-default-ipalloc] [--no-rewrite-hosts]
+                      [[--hostname-match <regexp>] [--hostname-replacement <replacement>]]
```
no need for the double brackets; the options can be set independently.

paulbellamy commented on proxy/proxy.go:
```diff
@@ -29,19 +31,22 @@ var (
-	dockerBridgeIP string
+	client              *docker.Client
+	dockerBridgeIP      string
+	hostnameMatchRegexp *regexp.Regexp
```
Where's `HostnameReplacement`? How does this compile?

paulbellamy commented on site/proxy.md:
```diff
@@ -130,6 +130,35 @@ automatically if it is running at the point when they are started -
 see the [weaveDNS usage](weavedns.html#usage) section for an in-depth
 explanation of the behaviour and how to control it.
 
+Typicaly, the proxy will pass on container names as-is to [weaveDNS](weavedns.html)
```
Typically

paulbellamy commented on site/proxy.md:
```diff
@@ -130,6 +130,35 @@ automatically if it is running at the point when they are started -
+
+For those situations, the proxy provides two flags: `--hostname-match <regexp>`
+and `--hostname-replacement <replacement>`. When launching a container, its name
+matched against regular expression `<regexp>`. Then, based on that match,
```
is matched

2opremio commented:
@tomwilkie It seems there are GCP quota problems in circleci https://circleci.com/gh/weaveworks/weave/1835

tomwilkie commented:
Yes I've just cleared out some stale VMs.  VMs aren't cleaned up if you
click cancel.

On Tue, Jul 14, 2015 at 4:15 PM, Alfonso Acosta <notifications@github.com>
wrote:

> @tomwilkie <https://github.com/tomwilkie> It seems there are GCP quota
> problems in circleci https://circleci.com/gh/weaveworks/weave/1835
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/weaveworks/weave/pull/1126#issuecomment-121279065>.
>

rade commented on weave:
```diff
@@ -37,6 +37,7 @@ weave launch-router [--password <password>] [--nickname <nickname>]
 weave launch-dns    [<addr>]
 weave launch-proxy  [-H <endpoint>] [--with-dns | --without-dns]
                       [--no-default-ipalloc] [--no-rewrite-hosts]
+                      [--hostname-match <regexp>] [--hostname-replacement <replacement>]
```
sorry, but `weave` usage needs to fit into 80 columns.

2opremio commented:
@rade Thanks. Out of curiosity, is there a reason why fix-commits are not squashed before merging? (Commits like ded3d6b40bdab5871b86731bd28cfdb0c3a67fe5 don't make any sense out of context) I added extra commits to make clear what was being changed, with the intention of squashing them before merging. Should I had kept amending instead?

rade commented:
I didn't realise you wanted to squash some commits. Better leave a note in the PR next time.

2opremio commented:
@rade Sorry, I was expecting the merger to check the commits before going ahead. This is probably not the forum for this discussion but, to avoid these situations, wouldn't it make sense to let the author merge the code himself after getting a "go ahead"?

rade commented:
> I was expecting the merger to check the commits before going ahead.

I did. I'm not bothered by the non-squashed commits; There is enough context to trace them back to this PR.

> This is probably not the forum for this discussion

Correct.

2opremio commented:
@rade ok, fair enough :)
